File: backend/bus/ts/producers/KafkaProducer.ts
// producers/kafka_producer.ts
import {
  Kafka,
  Producer,
  logLevel,
  IHeaders,
  CompressionTypes,
  Message,
} from "kafkajs";
import * as os from "node:os";

export type KafkaProducerOpts = {
  brokers?: string[];                 // ["host1:9092","host2:9092"]
  clientId?: string;                  // default: "<hostname>-<pid>"
  acks?: 0 | 1 | -1;                  // 0=none, 1=leader, -1=all (default -1)
  idempotent?: boolean;               // default true (requires acks=-1)
  compression?: "none" | "gzip" | "snappy" | "lz4" | "zstd";
  retries?: number;                   // producer-level retries (default 5)
  logLevel?: logLevel;                // default INFO
  allowAutoTopicCreation?: boolean;   // default false
};

const DEFAULT_HEADERS: IHeaders = {
  "content-type": "application/json",
  "x-producer": "kafkajs-json",
};

function encode(value: unknown): Buffer {
  return Buffer.from(JSON.stringify(value));
}

function toCompression(t: KafkaProducerOpts["compression"]) {
  switch ((t || "none").toLowerCase()) {
    case "gzip":   return CompressionTypes.GZIP;
    case "snappy": return CompressionTypes.Snappy;
    case "lz4":    return CompressionTypes.LZ4;
    case "zstd":   return CompressionTypes.ZSTD;
    default:       return CompressionTypes.None;
  }
}

export class KafkaJsonProducer {
  private kafka: Kafka;
  private producer: Producer;
  private readonly opts: Required<Omit<KafkaProducerOpts, "compression">> & {
    compression: KafkaProducerOpts["compression"];
  };

  constructor(opts: KafkaProducerOpts = {}) {
    const brokers = opts.brokers ?? (process.env.KAFKA_BROKERS || "localhost:9092").split(",");
    const clientId = opts.clientId ?? process.env.KAFKA_CLIENT_ID ?? `${os.hostname()}-${process.pid}`;
    const acks = opts.acks ?? -1;
    const idempotent = opts.idempotent ?? true;
    const retries = opts.retries ?? 5;
    const lvl = opts.logLevel ?? logLevel.INFO;

    this.kafka = new Kafka({ brokers, clientId, logLevel: lvl });

    this.producer = this.kafka.producer({
      allowAutoTopicCreation: opts.allowAutoTopicCreation ?? false,
      idempotent,
      maxInFlightRequests: idempotent ? 5 : 1,
      retry: { retries },
    });

    this.opts = {
      brokers,
      clientId,
      acks,
      idempotent,
      retries,
      logLevel: lvl,
      allowAutoTopicCreation: opts.allowAutoTopicCreation ?? false,
      compression: opts.compression ?? (process.env.KAFKA_COMPRESSION as any) ?? "none",
    };
  }

  /** Connect once at startup */
  async start(): Promise<void> {
    await this.producer.connect();
    process.once("SIGINT", () => void this.close());
    process.once("SIGTERM", () => void this.close());
  }

  /** Flush + disconnect safely */
async close(): Promise<void> {
  try {
    // flush() is deprecated / removed in newer kafkajs
    await this.producer.disconnect();
    console.log("[kafka-producer] disconnected cleanly");
  } catch (err) {
    console.error("[kafka-producer] error during disconnect:", err);
  }
}
  /** Publish one JSON record */
  async publish(params: {
    topic: string;
    value: unknown;                         // JSON-serializable
    key?: string | Buffer | null;
    headers?: IHeaders;
    partition?: number;
  }): Promise<void> {
    const { topic, value, key = null, headers = {}, partition } = params;

    const msg: Message = {
      value: encode(value),
      key: typeof key === "string" ? Buffer.from(key) : key ?? undefined,
      headers: { ...DEFAULT_HEADERS, ...headers },
      partition,
    };

    await this.producer.send({
      topic,
      acks: this.opts.acks,
      compression: toCompression(this.opts.compression),
      messages: [msg],
    });
  }

  /** Publish a batch (same topic) */
  async publishBatch(params: {
    topic: string;
    records: Array<{ value: unknown; key?: string | Buffer | null; headers?: IHeaders; partition?: number }>;
  }): Promise<void> {
    const { topic, records } = params;
    const messages: Message[] = records.map(r => ({
      value: encode(r.value),
      key: typeof r.key === "string" ? Buffer.from(r.key) : r.key ?? undefined,
      headers: { ...DEFAULT_HEADERS, ...(r.headers || {}) },
      partition: r.partition,
    }));

    await this.producer.send({
      topic,
      acks: this.opts.acks,
      compression: toCompression(this.opts.compression),
      messages,
    });
  }
}

/* ---------------- Example runner ---------------- */
if (require.main === module) {
  (async () => {
    const producer = new KafkaJsonProducer();
    await producer.start();

    // single
    await producer.publish({
      topic: process.env.KAFKA_TOPIC || "market.ticks.v1",
      key: "AAPL",
      value: { type: "tick", symbol: "AAPL", price: 191.25, ts: new Date().toISOString() },
      headers: { "x-event-type": "market" },
    });

    // batch
    await producer.publishBatch({
      topic: process.env.KAFKA_TOPIC || "market.ticks.v1",
      records: [
        { key: "MSFT",  value: { type: "tick", symbol: "MSFT",  price: 420.1, ts: new Date().toISOString() } },
        { key: "GOOGL", value: { type: "tick", symbol: "GOOGL", price: 130.0, ts: new Date().toISOString() } },
      ],
    });

    await producer.close();
    // eslint-disable-next-line no-process-exit
    process.exit(0);
  })().catch(async (e) => {
    console.error("[kafka-producer] fatal:", e);
    process.exit(1);
  });
}

File: backend/bus/python/producers/kafka_producer.py
# bus/python/producers/kafka_producer.py
from __future__ import annotations

import json
import os
import socket
import time
from typing import Any, Dict, Optional

from confluent_kafka import Producer # type: ignore


class KafkaProducerWrapper:
    """
    Simple Kafka producer wrapper.
    - Uses Avro/JSON (JSON here, but can be extended).
    - Adds metadata (env, host).
    - Auto-retry and flush support.
    """

    def __init__(
        self,
        brokers: Optional[str] = None,
        client_id: Optional[str] = None,
        acks: str = "all",
        linger_ms: int = 5,
    ):
        brokers = brokers or os.getenv("KAFKA_BROKERS", "localhost:9092")
        client_id = client_id or f"{socket.gethostname()}-{int(time.time())}"

        self._config: Dict[str, Any] = {
            "bootstrap.servers": brokers,
            "client.id": client_id,
            "acks": acks,
            "linger.ms": linger_ms,
        }
        self._producer = Producer(self._config)

    # ---------------------------
    # internal delivery report
    # ---------------------------
    def _delivery_report(self, err, msg):
        if err is not None:
            print(f"[KafkaProducer] Delivery failed for {msg.topic()} [{msg.partition()}]: {err}")
        else:
            print(
                f"[KafkaProducer] Delivered to {msg.topic()} [{msg.partition()}] @ offset {msg.offset()}"
            )

    # ---------------------------
    # publish JSON-serializable event
    # ---------------------------
    def publish(
        self,
        topic: str,
        event: Any,
        key: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        sync: bool = False,
    ) -> None:
        """
        Publish an event (dataclass or dict) to a Kafka topic.
        """
        if hasattr(event, "to_dict"):
            payload = event.to_dict()
        elif hasattr(event, "__dict__"):
            payload = event.__dict__
        elif isinstance(event, dict):
            payload = event
        else:
            raise TypeError("Event must be dataclass, dict, or have to_dict()")

        payload["env"] = os.getenv("APP_ENV", "prod")
        payload["emitted_at"] = int(time.time() * 1000)

        self._producer.produce(
            topic=topic,
            key=key,
            value=json.dumps(payload, separators=(",", ":")),
            headers=headers,
            callback=self._delivery_report,
        )

        if sync:
            self._producer.flush()

    def flush(self):
        """Flush outstanding messages."""
        self._producer.flush()


# =========================
# Example usage
# =========================
if __name__ == "__main__":
    from events.market import Tick

    producer = KafkaProducerWrapper()

    # Create a sample tick
    tick = Tick.create(symbol="AAPL", price=190.25, size=100, bid=190.20, ask=190.30, source="nasdaq")

    producer.publish(topic="market.ticks.v1", event=tick, key=tick.symbol, sync=True)